Sources/Apollo/HTTPNetworkTransport.swift

```diff
@@ -1,11 +1,78 @@
 import Foundation
 
+/// Methods which will be called prior to a request being sent to the server.
+public protocol HTTPNetworkTransportPreflightDelegate: class {
+  
+  /// Called when a request is about to send, to validate that it should be sent.
+  /// Good for early-exiting if your user is not logged in, for example.
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport which wants to send a request
+  ///   - request: The request, BEFORE it has been modified by `willSend`
+  /// - Returns: True if the request should proceed, false if not.
+  func networkTransport(_ networkTransport: HTTPNetworkTransport, shouldSend request: URLRequest) -> Bool
+  
+  /// Called when a request is about to send. Allows last minute modification of any properties on the request,
+  ///
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport which is about to send a request
+  ///   - request: The request, as an `inout` variable for modification
+  func networkTransport(_ networkTransport: HTTPNetworkTransport, willSend request: inout URLRequest)
+}
+
+// MARK: -
+
+/// Methods which will be called after some kind of response has been received to a `URLSessionTask`.
+public protocol HTTPNetworkTransportTaskCompletedDelegate {
+  
+  /// A callback to allow hooking in URL session responses for things like logging and examining headers.
+  /// NOTE: This will call back on whatever thread the URL session calls back on, which is never the main thread. Call `DispatchQueue.main.async` before touching your UI!
+  ///
+  /// - Parameters:
+  ///   - networkTransport: The network transport that completed a task
+  ///   - request: The request which was completed by the task
+  ///   - data: [optional] Any data received. Passed through from `URLSession`.
+  ///   - response: [optional] Any response received. Passed through from `URLSession`.
+  ///   - error: [optional] Any error received. Passed through from `URLSession`.
+  func networkTransport(_ networkTransport: HTTPNetworkTransport,
```

Comment:
For consistency with platform methods maybe something like `networkTransport:didCompleteTaskForRequest:withData:response:error`? So `didComplete` instead of `completed`, and I'm not sure whether the `raw` part is necessary? Alternatively, we could leave out the `task` altogether, and make it something like `networkTransport:didCompleteRequest:withData:response:error`.

Reply:
Agree on `didComplete` vs `completed`, will update. 

The `Raw` is to indicate that this is unprocessed data (and that the consumer shouldn't necessarily be trying to process it themselves). I started with `didCompleteRequest` but I went to the bit with `Raw` because I didn't want it to be confused by users with the actual completion closure.